### ISSUE TITLE
Test and refine guided-generate functionality across parameter space and alternative approaches

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/TinyDiffusionModels#1]
## Overview

The current guided-generate functionality (`src/shakespeare.py:176-204`) combines autoregressive LM logits with diffusion-generated embeddings, but needs systematic testing and refinement to improve text quality and coherence.

## Current Implementation Analysis

**Core Function**: `guided_generate()` mixes AR and diffusion logits:
```python
mixed_logits = (1 - alpha) * ar_logits + alpha * diff_logits
```

**Key Parameters**:
- `alpha`: Controls AR vs diffusion mixing (currently defaults to 0.3)
- `temperature`: Controls sampling randomness (defaults to 1.0)
- `max_len`: Generation length (currently uses diffusion sequence length)

**Current Issues**:
- Generated text often lacks coherence (see `samples/guided_sample_*.txt`)
- Fixed greedy decoding limits diversity
- Single alpha value may not be optimal across all positions
- No systematic evaluation of parameter combinations

### 1. Parameter Space Exploration

**Alpha Sweep**:
- Test values: [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
- Generate 10 samples per alpha value
- Evaluate coherence, diversity, and Shakespeare-likeness

**Temperature Sweep**:
- Test values: [0.5, 0.7, 1.0, 1.2, 1.5]
- Combined with optimal alpha values from above
- Assess impact on output randomness vs quality

**Dynamic Alpha**:
- Position-dependent alpha: `alpha(pos) = base_alpha * decay_factor^pos`
- Investigate if early AR guidance + late diffusion works better

### 2. Alternative Mixing Strategies

**Probability-Based Mixing**:
```python
# Instead of logit mixing
ar_probs = F.softmax(ar_logits, dim=-1)
diff_probs = F.softmax(diff_logits, dim=-1)
mixed_probs = (1 - alpha) * ar_probs + alpha * diff_probs
```

**Top-K Filtering**:
- Apply top-k filtering to AR logits before mixing
- Test k values: [10, 20, 50, 100]

**Entropy-Based Adaptive Mixing**:
- Use AR prediction confidence to dynamically adjust alpha
- High confidence → lower alpha (trust AR)
- Low confidence → higher alpha (rely on diffusion)

**Nucleus (top-p) Sampling**:
- Replace greedy decoding with nucleus sampling
- Test p values: [0.8, 0.9, 0.95]

### 3. Embedding Space Refinements

**Embedding Interpolation**:
- Instead of cosine similarity, use learned projection
- Train small MLP: `diffusion_embedding → vocabulary_logits`

**Multi-Scale Diffusion**:
- Generate embeddings at multiple scales
- Combine local (token-level) and global (sequence-level) guidance

**Positional Embedding Injection**:
- Add positional information to diffusion embeddings
- Test if position-aware diffusion improves coherence

### 4. Evaluation Metrics

**Automatic Metrics**:
- Perplexity (using base LM)
- BLEU score against Shakespeare corpus
- Self-BLEU for diversity measurement
- Embedding similarity to Shakespeare style

**Human-Interpretable Metrics**:
- Coherence score (1-5 scale)
- Shakespeare-likeness (1-5 scale)
- Grammatical correctness
- Vocabulary richness (unique token ratio)

### 5. Implementation Plan

**Phase 1: Parameter Grid Search**
- [ ] Implement parameter sweep framework
- [ ] Generate samples across alpha/temperature grid
- [ ] Basic quality assessment

**Phase 2: Alternative Strategies**
- [ ] Implement probability-based mixing
- [ ] Add top-k/nucleus sampling options
- [ ] Test dynamic alpha strategies

**Phase 3: Advanced Techniques**
- [ ] Learned embedding projection
- [ ] Multi-scale diffusion guidance
- [ ] Positional embedding enhancement

**Phase 4: Evaluation & Optimization**
- [ ] Comprehensive metric evaluation
- [ ] Statistical significance testing
- [ ] Select optimal configuration

### 6. Technical Implementation

**Configuration Structure**:
```python
@dataclass
class GuidedGenerationConfig:
    alpha: float = 0.3
    temperature: float = 1.0
    sampling_strategy: str = "greedy"  # greedy, top_k, nucleus
    top_k: int = 50
    top_p: float = 0.9
    dynamic_alpha: bool = False
    alpha_decay: float = 0.95
    use_probability_mixing: bool = False
```

**Evaluation Pipeline**:
- Automated generation across parameter space
- Batch evaluation with multiple metrics
- Statistical analysis and visualization
- Best configuration selection

### 7. Expected Outcomes

- Identify optimal parameter combinations for different use cases
- Implement robust alternative mixing strategies
- Establish evaluation framework for future improvements
- Document best practices for guided generation

### 8. Success Criteria

- Generated text shows improved coherence over current implementation
- Clear parameter recommendations for different quality/diversity trade-offs
- Reproducible evaluation framework
- At least 2-3 alternative approaches that outperform current method

---

**Priority**: High
**Complexity**: Medium-High
**Dependencies**: Existing diffusion model checkpoints
**Timeline**: 2-3 weeks for comprehensive testing
